### PR TITLE
Support OCaml for Windows +mingw{32,64} compiler variant

### PR DESCRIPTION
**Pull request by MisterDA**

### CODE REVIEW

File: ocaml_version.ml [hunk 1]
```diff
@@ -364,6 +371,10 @@ module Configure_options = struct
       | `Frame_pointer -> "--enable-frame-pointers"
       | `No_naked_pointers -> "--disable-naked-pointers"
       | `Disable_flat_float_array -> "--disable-flat-float-array"
+      | `Mingw `X86_64 ->
+         "--host=x86_64-w64-mingw32 CC=x86_64-w64-mingw32-gcc CPP=x86_64-w64-mingw32-cpp"
```
Comment (dra27) [inline ocaml_version.ml:375]: `CC` and `CPP` should not need overriding?

**Review by dra27 — COMMENTED** [review]
There's a few problems here:

- Upstream opam-repository will never have these variants, so we need to consider how that changes in the future
- The point with the `Configure_options` at the moment is that you can combine them, which doesn't make sense for the `` `Mingw`` option (e.g. ````[`Mingw `I386; `Mingw `X86_64]````)
- These aren't really variants in the normal way - i.e. these don't map to the new `ocaml-option-*` packages upstream

I think the important function which needs to work at the moment is to be able to get `Opam.package`, `Opam.additional_packages` and `Opam.variant_switch` to return the correct values, right?

I think this should be doable by introducing `?host` to these functions where host will be an `arch * platform` pair. So you'd then expect `Opam.V2.package ~host:(`X86_64, `Mingw) Releases.v4_12` to return `("ocaml-variants", "4.12.0+mingw64c")`.

That way (assuming it does indeed work!) when the Windows switches are upstreamed to opam-repository, we simply change the package options which get returned - in this case, switching to `("ocaml-base-compiler", "4.12.0")` and returning something like `"c-compiler-mingw"` in `additional_packages`

File: ocaml_version.ml (hunk 1 above)
Comment (MisterDA) [reply ocaml_version.ml:375]: I wasn't sure, I saw the build script in opam-repository-mingw did override them, so I reluctantly added the two vars.

**Comment by MisterDA** [comment]
I see, and I agree -- I'll try your proposal
Should we use `+mingwXXc` as-in _precompiled_ or `+mingwXX` and let ocaml be compiled when creating the switch?

File: ocaml_version.ml (hunk 1 above)
Comment (dra27) [reply ocaml_version.ml:375]: Yeah, that override drives me up the wall developing (I've been stung by it several times on machines which happen to have it installed!). It definitely won't make it upstream...

**Comment by MisterDA** [comment]
I'm closing this PR for now, the `msvcXX`/`mingwXX` variants from fdopen packages can be defined in ocaml-dockerfile or in the ocurrent pipeline. It's easier, and we'll see what needs to be changed here when we'll get to opam 2.2.

**Pull request closed without merging**